### PR TITLE
New Integration Test Added for Stack Analyses Dynamic Manifests

### PR DESCRIPTION
**Pull request by deepak1725**
# Description

New Integration Test Added for Stack Analyses Dynamic Manifests

## Type of change

Please delete options that are not relevant.
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by deepak1725** [comment]
> Could you please point me to the code that generates dynamic manifests and stores them on S3?

@tisnik PR is still Open https://github.com/fabric8-analytics/f8a-stacks-report/pull/120

**Comment by tisnik** [comment]
Ah @deepak1725 without that change, the e2e will likely fail. Please mark this PR with WIP:

File: integration-tests/features/steps/stack_analysis.py [hunk 1]
```diff
@@ -800,3 +809,50 @@ def validate_topic_list(context, key):
...
+
+@when('I tried to fetch dynamic manifests from s3')
+def dynamic_manifest_file(context):
+    """Integration for dynamically generated manifests files."""
+    response, status = context.s3interface.get_object_from_s3(bucket_name=context.reports_bucket)
+    assert status == 200, response
+
+
+@then('I should be able to validate them')
+def validate_dynamic_manifest_file(context):
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: It is not validation step, right? It just retrieves data from S3 w/o any validation

File: integration-tests/features/steps/stack_analysis.py [hunk 2]
```diff
@@ -800,3 +809,50 @@ def validate_topic_list(context, key):
...
+        ('pylist.json', validate_file_pypi)
+    ]
+    for file, func in files_to_validate:
+        with open('data/dynamic_manifests/'+file) as fp:
+            contents = json.load(fp)
+        contents = random.sample(contents,1)
+        func(contents[0], file, file_save_path)
+
+
+def validate_file_npm(contents,file_name, file_save_path):
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: dtto it is not validation AFAIK

File: integration-tests/features/steps/stack_analysis.py [hunk 3]
```diff
@@ -800,3 +809,50 @@ def validate_topic_list(context, key):
...
+    """Generate file for npm."""
+    data = {'dependencies':{}}
+    with open(file_save_path+file_name, 'w') as manifest:
+        data['dependencies'] = contents.pop('dependencies')
+        json.dump(data, manifest)
+
+
+def validate_file_pypi(contents, file_name, file_save_path):
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: dtto

File: integration-tests/features/steps/stack_analysis.py [hunk 4]
```diff
@@ -800,3 +809,50 @@ def validate_topic_list(context, key):
...
+    """Generate file for pypi."""
+    with open(file_save_path+file_name, 'w') as manifest:
+        json.dump(contents[0], manifest)
+
+
+def validate_file_maven(contents, file_name, file_save_path):
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: dtto

File: integration-tests/features/steps/stack_analysis.py (hunk 1 above)
Comment (deepak1725) [reply integration-tests/features/steps/stack_analysis.py]: @tisnik .. List of Manifests are received from S3, So, this function randomly selects one stack of them and generate a valid manifest out of it.

**Comment by deepak1725** [comment]
Relevant PR's:
https://github.com/fabric8-analytics/f8a-stacks-report/pull/126
https://github.com/fabric8-analytics/f8a-stacks-report/pull/129
https://github.com/fabric8-analytics/f8a-stacks-report/pull/133

**Comment by deepak1725** [comment]
@tisnik All relevant PR's are merged and closed and Manifests files are avaible to process in S3.

File: integration-tests/MAINTAINERS [hunk 5]
```diff
@@ -1,2 +1,3 @@
 Pavel Tisnovsky <ptisnovs@redhat.com>
 Michal Srb <msrb@redhat.com>
+Deepak Sharma <deepshar@redhat.com>
```
Comment (tisnik) [inline integration-tests/MAINTAINERS]: you might remove most maintainers who no longer work on this project

File: integration-tests/features/src/s3interface.py [hunk 6]
```diff
@@ -139,3 +141,23 @@ def read_object_metadata(self, bucket_name, key, attribute):
...
+        for obj in files_to_download:
+            save_path = base_url + path + obj
+            try:
+                filedata = urlrequest.urlopen(url=save_path).read()
+                with open(path + obj, 'wb') as f:
+                    f.write(filedata)
+            except botocore.exceptions.ClientError:
+                print("The object does not exist.", obj)
+                return 'error in file fetching from s3', 400
+        return 'success', 200
```
Comment (tisnik) [inline integration-tests/features/src/s3interface.py]: why do you return something like HTTP status code. Is not better to raise exception or (at least) return flag whether the operation is ok or not. Is this function copied from another source code, right?

File: integration-tests/features/stack_analyses.feature [hunk 7]
```diff
@@ -134,3 +134,17 @@ Feature: Analyse API functionality check
...
+    Then I should get the proper authorization token
+    When I wait 10 seconds
+    When I tried to fetch dynamic manifests from s3
+    Then I should be able to validate them
+    When I send NPM package manifest valid_manifests/npmlist.json to stack analysis version 3 with authorization token
+    Then I should get 200 status code
+    When I send Python package manifest valid_manifests/pylist.json to stack analysis version 3 with authorization token
+    Then I should get 200 status code
+    When I send new Maven package manifest valid_manifests/dependencies.txt to stack analysis version 3 with authorization token
+    Then I should get 200 status code
```
Comment (tisnik) [inline integration-tests/features/stack_analyses.feature]: might be good to check the content of stack analysis

File: integration-tests/features/steps/stack_analysis.py [hunk 8]
```diff
@@ -800,3 +810,50 @@ def validate_topic_list(context, key):
...
+
+def validate_file_pypi(contents, file_name, file_save_path):
+    """Generate file for pypi."""
+    with open(file_save_path + file_name, 'w') as manifest:
+        json.dump(contents[0], manifest)
+
+
+def validate_file_maven(contents, file_name, file_save_path):
+    """Generate file for maven."""
+    with open(file_save_path + 'dependencies.txt', 'w') as manifest:
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: you can use pathlib library there

File: integration-tests/features/steps/stack_analysis.py [hunk 9]
```diff
@@ -800,3 +810,50 @@ def validate_topic_list(context, key):
...
+    """Generate file for npm."""
+    data = {'dependencies': {}}
+    with open(file_save_path + file_name, 'w') as manifest:
+        data['dependencies'] = contents.pop('dependencies')
+        json.dump(data, manifest)
+
+
+def validate_file_pypi(contents, file_name, file_save_path):
+    """Generate file for pypi."""
+    with open(file_save_path + file_name, 'w') as manifest:
```
Comment (tisnik) [inline integration-tests/features/steps/stack_analysis.py]: you can use pathlib there

File: integration-tests/features/src/s3interface.py (hunk 6 above)
Comment (deepak1725) [reply integration-tests/features/src/s3interface.py]: @tisnik If we can see caller func. https://github.com/deepak1725/fabric8-analytics-common/blob/6ff59c8ab7a0ea8383c974cc53faf75fbb9f7d27/integration-tests/features/steps/stack_analysis.py#L818

Here assert will trigger only when format remains the same in success and fail scenerios.

Also i am Okay to update it to raise Exception here. Please suggest

File: integration-tests/features/stack_analyses.feature (hunk 7 above)
Comment (deepak1725) [reply integration-tests/features/stack_analyses.feature]: File contents are already validated in stack analyses API,
i feel its kind of redundant here, as
Tests will Fail in either case:
a) We write a function to check file contents and contents are not valid.
b) We send the file as it is for Stack Analyses and hence server will throw an Error if file is not valid.
Please suggest.

File: integration-tests/features/steps/stack_analysis.py (hunk 9 above)
Comment (deepak1725) [reply integration-tests/features/steps/stack_analysis.py]: IMHO facing Issue with Python3.4
Failed Job: https://ci.centos.org/job/devtools-e2e-fabric8-analytics/1423/
More Info: https://stackoverflow.com/questions/42694112/when-using-pathlib-getting-error-typeerror-invalid-file-posixpathexample-t

File: integration-tests/features/steps/stack_analysis.py (hunk 8 above)
Comment (deepak1725) [reply integration-tests/features/steps/stack_analysis.py]: ditto https://github.com/fabric8-analytics/fabric8-analytics-common/pull/1157#discussion_r349024042

**Comment by deepak1725** [comment]
[test]

File: integration-tests/features/stack_analyses.feature (hunk 7 above)
Comment (tisnik) [reply integration-tests/features/stack_analyses.feature]: I meant that we should check the result of stack analysis, ie. the schema, whether all packages are mentioned etc. etc.

Comment (deepak1725) [reply integration-tests/features/stack_analyses.feature]: @tisnik Added more scnerios. Please re-review

**Comment by deepak1725** [comment]
[test]

**Pull request merged**